File: internal/awg/vpnuri/vpnuri.go
// Copyright (c) 2025 LucX-UI Project.
// Licensed under the PolyForm Noncommercial License 1.0.0.
// LucX-UI Component. Free for personal and educational use.
// Commercial use (including VPN resale) requires explicit written permission from the author.
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0

// Package vpnuri encodes AmneziaVPN vpn:// share links from a WireGuard/AWG
// client .conf. Format matches amnezia-client ExportController:
//
//	vpn:// + Base64URL( qCompress(JSON) )
//
// where qCompress is Qt's wrapper: 4-byte big-endian uncompressed length +
// raw DEFLATE (zlib) body. The JSON is the official Amnezia container
// (containers[].awg.last_config) so NekoBox+ and awg-manager parse it.
// last_config is a JSON string matching Amnezia AwgClientConfig (config plus
// client_priv_key / server_pub_key / Jc / …). processAmneziaConfig skips
// parsing when last_config is already JSON; without the structured keys the
// app imports the container but the handshake dies immediately.
package vpnuri

import (
	"bytes"
	"compress/zlib"
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

const (
	containerName = "amnezia-awg"
	protoName     = "awg"
)

// EncodeConf builds a vpn:// URI from an awg-quick client .conf body.
func EncodeConf(conf string) (string, error) {
	conf = strings.TrimSpace(conf)
	if conf == "" {
		return "", fmt.Errorf("vpnuri: empty conf")
	}
	if !strings.Contains(conf, "[Interface]") || !strings.Contains(conf, "[Peer]") {
		return "", fmt.Errorf("vpnuri: conf missing [Interface]/[Peer]")
	}
	payload, err := json.Marshal(amneziaEnvelope(conf))
	if err != nil {
		return "", err
	}
	compressed, err := qCompress(payload)
	if err != nil {
		return "", err
	}
	return "vpn://" + base64.RawURLEncoding.EncodeToString(compressed), nil
}

func amneziaEnvelope(conf string) map[string]any {
	meta := parseConf(conf)
	innerJSON, _ := json.Marshal(meta.inner)
	awg := map[string]any{
		"last_config":        string(innerJSON),
		"isThirdPartyConfig": true,
		"transport_proto":    "udp",
	}
	if meta.port > 0 {
		awg["port"] = strconv.Itoa(meta.port)
	}
	// AmneziaVPN picks the client generation off protocol_version; without it
	// older app builds fall back to v1 obfuscation and the AWG3 handshake dies.
	if pv := protocolVersion(meta.inner); pv != "" {
		awg["protocol_version"] = pv
	}
	env := map[string]any{
		"defaultContainer": containerName,
		"containers": []any{
			map[string]any{
				"container": containerName,
				protoName:   awg,
			},
		},
	}
	if meta.host != "" {
		env["hostName"] = meta.host
	}
	if meta.desc != "" {
		env["description"] = meta.desc
	}
	if meta.dns1 != "" {
		env["dns1"] = meta.dns1
	}
	if meta.dns2 != "" {
		env["dns2"] = meta.dns2
	}
	return env
}

// awgLastConfigKeys are Amnezia last_config field names that match .conf keys
// 1:1 (extractWireGuardConfig copies them as-is).
var awgLastConfigKeys = map[string]string{
	"jc":                     "Jc",
	"jmin":                   "Jmin",
	"jmax":                   "Jmax",
	"s1":                     "S1",
	"s2":                     "S2",
	"s3":                     "S3",
	"s4":                     "S4",
	"h1":                     "H1",
	"h2":                     "H2",
	"h3":                     "H3",
	"h4":                     "H4",
	"i1":                     "I1",
	"i2":                     "I2",
	"i3":                     "I3",
	"i4":                     "I4",
	"i5":                     "I5",
	"headerprotectionkey":    "HeaderProtectionKey",
	"contentpaddingaddition": "ContentPaddingAddition",
	"rekeyaftertime":         "RekeyAfterTime",
	"rekeytimeout":           "RekeyTimeout",
	"rejectaftertime":        "RejectAfterTime",
	"keepalivetimeout":       "KeepaliveTimeout",
	"maxhandshakeattempts":   "MaxHandshakeAttempts",
	"randomtrailers":         "RandomTrailers",
	"disablecookies":         "DisableCookies",
}

type confMeta struct {
	host, dns1, dns2, desc string
	port                   int
	inner                  map[string]any
}

func parseConf(conf string) confMeta {
	meta := confMeta{inner: map[string]any{"config": conf}}
	for _, raw := range strings.Split(conf, "\n") {
		line := strings.TrimSpace(raw)
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "#") {
			if meta.desc == "" {
				meta.desc = strings.TrimSpace(strings.TrimPrefix(line, "#"))
			}
			continue
		}
		if strings.HasPrefix(line, "[") {
			continue
		}
		key, val, ok := strings.Cut(line, "=")
		if !ok {
			continue
		}
		key = strings.TrimSpace(key)
		val = strings.TrimSpace(val)
		if val == "" {
			continue
		}
		switch strings.ToLower(key) {
		case "endpoint":
			if meta.host == "" {
				meta.host, meta.port = splitHostPort(val)
				if meta.host != "" {
					meta.inner["hostName"] = meta.host
				}
				if meta.port > 0 {
					meta.inner["port"] = meta.port
				}
			}
		case "dns":
			parts := strings.Split(val, ",")
			if meta.dns1 == "" && len(parts) > 0 {
				meta.dns1 = strings.TrimSpace(parts[0])
			}
			if meta.dns2 == "" && len(parts) > 1 {
				meta.dns2 = strings.TrimSpace(parts[1])
			}
		case "privatekey":
			meta.inner["client_priv_key"] = val
		case "address":
			meta.inner["client_ip"] = val
		case "publickey":
			meta.inner["server_pub_key"] = val
		case "presharedkey":
			meta.inner["psk_key"] = val
		case "mtu":
			meta.inner["mtu"] = val
		case "persistentkeepalive":
			meta.inner["persistent_keep_alive"] = val
		case "allowedips":
			var ips []string
			for _, p := range strings.Split(val, ",") {
				if ip := strings.TrimSpace(p); ip != "" {
					ips = append(ips, ip)
				}
			}
			if len(ips) > 0 {
				meta.inner["allowed_ips"] = ips
			}
		default:
			if jsonKey, ok := awgLastConfigKeys[strings.ToLower(key)]; ok {
				meta.inner[jsonKey] = val
			}
		}
	}
	return meta
}

// awg3ProtocolVersionKeys mark an AWG 3.0 config: any of them present forces
// protocol_version "3" in the envelope.
var awg3ProtocolVersionKeys = []string{
	"HeaderProtectionKey",
	"ContentPaddingAddition",
	"RekeyAfterTime",
	"RekeyTimeout",
	"RejectAfterTime",
	"KeepaliveTimeout",
	"MaxHandshakeAttempts",
}

func protocolVersion(inner map[string]any) string {
	for _, key := range awg3ProtocolVersionKeys {
		if s, ok := inner[key].(string); ok && strings.TrimSpace(s) != "" {
			return "3"
		}
	}
	if atoiPositive(inner, "S3") || atoiPositive(inner, "S4") {
		return "2"
	}
	for _, key := range []string{"H1", "H2", "H3", "H4"} {
		if s, ok := inner[key].(string); ok && strings.Contains(s, "-") {
			return "2"
		}
	}
	if s, ok := inner["I1"].(string); ok && strings.TrimSpace(s) != "" {
		return "2"
	}
	return ""
}

func atoiPositive(inner map[string]any, key string) bool {
	s, ok := inner[key].(string)
	if !ok {
		return false
	}
	n, err := strconv.Atoi(strings.TrimSpace(s))
	return err == nil && n > 0
}

func splitHostPort(s string) (string, int) {
	if strings.HasPrefix(s, "[") {
		end := strings.LastIndex(s, "]")
		if end < 0 {
			return s, 0
		}
		host := s[1:end]
		rest := s[end+1:]
		if !strings.HasPrefix(rest, ":") {
			return host, 0
		}
		p, err := strconv.Atoi(rest[1:])
		if err != nil {
			return host, 0
		}
		return host, p
	}
	i := strings.LastIndex(s, ":")
	if i < 0 {
		return s, 0
	}
	p, err := strconv.Atoi(s[i+1:])
	if err != nil {
		return s, 0
	}
	return s[:i], p
}

func qCompress(data []byte) ([]byte, error) {
	var buf bytes.Buffer
	var size [4]byte
	binary.BigEndian.PutUint32(size[:], uint32(len(data)))
	buf.Write(size[:])
	w := zlib.NewWriter(&buf)
	if _, err := w.Write(data); err != nil {
		_ = w.Close()
		return nil, err
	}
	if err := w.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// Decode returns uncompressed payload bytes (JSON envelope or legacy raw .conf).
func Decode(uri string) ([]byte, error) {
	s := strings.TrimSpace(uri)
	s = strings.TrimPrefix(s, "vpn://")
	raw, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		raw, err = base64.URLEncoding.DecodeString(s)
		if err != nil {
			return nil, err
		}
	}
	if len(raw) < 4 {
		return nil, fmt.Errorf("vpnuri: truncated payload")
	}
	r, err := zlib.NewReader(bytes.NewReader(raw[4:]))
	if err != nil {
		return nil, err
	}
	defer r.Close()
	var out bytes.Buffer
	if _, err := out.ReadFrom(r); err != nil {
		return nil, err
	}
	return out.Bytes(), nil
}

// ConfFromPayload extracts the awg-quick .conf from a Decode() result.
// Accepts the official JSON container and the legacy raw-conf payload.
func ConfFromPayload(payload []byte) (string, error) {
	trimmed := strings.TrimSpace(string(payload))
	if trimmed != "" && trimmed[0] != '{' && strings.Contains(trimmed, "[Interface]") && strings.Contains(trimmed, "[Peer]") {
		return trimmed, nil
	}
	var env map[string]any
	if err := json.Unmarshal(payload, &env); err != nil {
		return "", fmt.Errorf("vpnuri: not JSON and not .conf")
	}
	containers, _ := env["containers"].([]any)
	if len(containers) == 0 {
		return "", fmt.Errorf("vpnuri: no containers")
	}
	last, _ := containers[len(containers)-1].(map[string]any)
	if last == nil {
		return "", fmt.Errorf("vpnuri: bad container")
	}
	proto, _ := last["awg"].(map[string]any)
	if proto == nil {
		proto, _ = last["wireguard"].(map[string]any)
	}
	if proto == nil {
		return "", fmt.Errorf("vpnuri: no awg/wireguard")
	}
	switch lc := proto["last_config"].(type) {
	case string:
		var inner map[string]any
		if err := json.Unmarshal([]byte(lc), &inner); err == nil {
			if c, ok := inner["config"].(string); ok && strings.Contains(c, "[Interface]") {
				return strings.TrimSpace(c), nil
			}
		}
		if strings.Contains(lc, "[Interface]") {
			return strings.TrimSpace(lc), nil
		}
	case map[string]any:
		if c, ok := lc["config"].(string); ok && strings.Contains(c, "[Interface]") {
			return strings.TrimSpace(c), nil
		}
	}
	return "", fmt.Errorf("vpnuri: last_config has no .conf")
}

File: frontend/src/lib/awg/vpnuri.ts
// Copyright (c) 2025 LucX-UI Project.
// Licensed under the PolyForm Noncommercial License 1.0.0.
// LucX-UI Component. Free for personal and educational use.
// Commercial use (including VPN resale) requires explicit written permission from the author.
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0

export function isQCompress(bytes: Uint8Array): boolean {
  if (bytes.length < 6) return false;
  return bytes[4] === 0x78 && [0x01, 0x5e, 0x9c, 0xda].includes(bytes[5]);
}

export function bytesFromBase64Url(value: string): Uint8Array {
  const b64 = value.replace(/-/g, '+').replace(/_/g, '/');
  const padded = b64 + '='.repeat((4 - (b64.length % 4)) % 4);
  const binary = atob(padded);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

function toBase64Url(bytes: Uint8Array): string {
  let binary = '';
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

// Qt qCompress framing: 4-byte big-endian uncompressed length + zlib stream.
// The deflate body uses stored (uncompressed) blocks — the most basic valid
// zlib form, accepted by Qt's qUncompress and every inflate implementation,
// which keeps this encoder synchronous (CompressionStream is async-only).
function qCompress(data: Uint8Array): Uint8Array {
  const blocks: number[] = [];
  let offset = 0;
  do {
    const len = Math.min(data.length - offset, 65535);
    const final = offset + len >= data.length;
    const nlen = len ^ 0xffff;
    blocks.push(
      final ? 0x01 : 0x00,
      len & 0xff,
      (len >> 8) & 0xff,
      nlen & 0xff,
      (nlen >> 8) & 0xff,
    );
    for (let i = 0; i < len; i++) blocks.push(data[offset + i]);
    offset += len;
  } while (offset < data.length);

  let a = 1;
  let b = 0;
  for (const byte of data) {
    a = (a + byte) % 65521;
    b = (b + a) % 65521;
  }
  const adler = ((b << 16) | a) >>> 0;

  const out = new Uint8Array(4 + 2 + blocks.length + 4);
  const view = new DataView(out.buffer);
  view.setUint32(0, data.length, false);
  out[4] = 0x78;
  out[5] = 0x01;
  out.set(blocks, 6);
  view.setUint32(out.length - 4, adler, false);
  return out;
}

// Amnezia last_config field names that match .conf keys 1:1 (same table as
// internal/awg/vpnuri on the Go side and Amnezia's own awgProtocolKeys).
const AWG_CONF_KEYS: Record<string, string> = {
  jc: 'Jc',
  jmin: 'Jmin',
  jmax: 'Jmax',
  s1: 'S1',
  s2: 'S2',
  s3: 'S3',
  s4: 'S4',
  h1: 'H1',
  h2: 'H2',
  h3: 'H3',
  h4: 'H4',
  i1: 'I1',
  i2: 'I2',
  i3: 'I3',
  i4: 'I4',
  i5: 'I5',
  headerprotectionkey: 'HeaderProtectionKey',
  contentpaddingaddition: 'ContentPaddingAddition',
  rekeyaftertime: 'RekeyAfterTime',
  rekeytimeout: 'RekeyTimeout',
  rejectaftertime: 'RejectAfterTime',
  keepalivetimeout: 'KeepaliveTimeout',
  maxhandshakeattempts: 'MaxHandshakeAttempts',
  randomtrailers: 'RandomTrailers',
  disablecookies: 'DisableCookies',
};

interface ConfMeta {
  host: string;
  port: number;
  dns1: string;
  dns2: string;
  desc: string;
  inner: Record<string, unknown>;
}

function splitHostPort(s: string): { host: string; port: number } {
  if (s.startsWith('[')) {
    const end = s.lastIndexOf(']');
    if (end < 0) return { host: s, port: 0 };
    const rest = s.slice(end + 1);
    if (!rest.startsWith(':')) return { host: s.slice(1, end), port: 0 };
    const p = Number(rest.slice(1));
    return { host: s.slice(1, end), port: Number.isInteger(p) && p > 0 ? p : 0 };
  }
  const i = s.lastIndexOf(':');
  if (i < 0) return { host: s, port: 0 };
  const p = Number(s.slice(i + 1));
  return { host: s.slice(0, i), port: Number.isInteger(p) && p > 0 ? p : 0 };
}

function parseConf(conf: string): ConfMeta {
  const meta: ConfMeta = {
    host: '',
    port: 0,
    dns1: '',
    dns2: '',
    desc: '',
    inner: { config: conf },
  };
  for (const raw of conf.split('\n')) {
    const line = raw.trim();
    if (!line) continue;
    if (line.startsWith('#')) {
      if (!meta.desc) meta.desc = line.slice(1).trim();
      continue;
    }
    if (line.startsWith('[')) continue;
    const eq = line.indexOf('=');
    if (eq <= 0) continue;
    const key = line.slice(0, eq).trim();
    const val = line.slice(eq + 1).trim();
    if (!val) continue;
    switch (key.toLowerCase()) {
      case 'endpoint':
        if (!meta.host) {
          const { host, port } = splitHostPort(val);
          meta.host = host;
          meta.port = port;
          if (host) meta.inner.hostName = host;
          if (port > 0) meta.inner.port = port;
        }
        break;
      case 'dns': {
        const parts = val.split(',').map((p) => p.trim());
        if (!meta.dns1 && parts.length > 0) meta.dns1 = parts[0];
        if (!meta.dns2 && parts.length > 1) meta.dns2 = parts[1];
        break;
      }
      case 'privatekey':
        meta.inner.client_priv_key = val;
        break;
      case 'address':
        meta.inner.client_ip = val;
        break;
      case 'publickey':
        meta.inner.server_pub_key = val;
        break;
      case 'presharedkey':
        meta.inner.psk_key = val;
        break;
      case 'mtu':
        meta.inner.mtu = val;
        break;
      case 'persistentkeepalive':
        meta.inner.persistent_keep_alive = val;
        break;
      case 'allowedips': {
        const ips = val
          .split(',')
          .map((p) => p.trim())
          .filter((p) => p !== '');
        if (ips.length > 0) meta.inner.allowed_ips = ips;
        break;
      }
      default: {
        const jsonKey = AWG_CONF_KEYS[key.toLowerCase()];
        if (jsonKey) meta.inner[jsonKey] = val;
      }
    }
  }
  return meta;
}

function nonEmpty(inner: Record<string, unknown>, key: string): boolean {
  const v = inner[key];
  return typeof v === 'string' && v.trim() !== '';
}

function atoiPositive(inner: Record<string, unknown>, key: string): boolean {
  const v = inner[key];
  if (typeof v !== 'string') return false;
  const n = Number(v.trim());
  return Number.isInteger(n) && n > 0;
}

// AmneziaVPN picks the protocol generation off protocol_version; without it
// older app builds fall back to v1 obfuscation and the AWG3 handshake dies.
function protocolVersion(inner: Record<string, unknown>): string {
  const awg3 = [
    'HeaderProtectionKey',
    'ContentPaddingAddition',
    'RekeyAfterTime',
    'RekeyTimeout',
    'RejectAfterTime',
    'KeepaliveTimeout',
    'MaxHandshakeAttempts',
  ];
  if (awg3.some((k) => nonEmpty(inner, k))) return '3';
  if (atoiPositive(inner, 'S3') || atoiPositive(inner, 'S4')) return '2';
  if (['H1', 'H2', 'H3', 'H4'].some((k) => nonEmpty(inner, k) && String(inner[k]).includes('-'))) {
    return '2';
  }
  if (nonEmpty(inner, 'I1')) return '2';
  return '';
}

// Synchronous inflate for the stored-block zlib streams vpnUriFromConf emits.
// Real deflate streams (the Go sub service uses zlib compression) return null;
// callers that must decode those use the async DecompressionStream path.
export function inflateStored(bytes: Uint8Array): Uint8Array | null {
  if (bytes.length < 6 || bytes[4] !== 0x78) return null;
  let i = 6;
  const out: number[] = [];
  for (;;) {
    if (i + 5 > bytes.length) return null;
    const header = bytes[i];
    if ((header & 0x06) !== 0) return null;
    const len = bytes[i + 1] | (bytes[i + 2] << 8);
    const nlen = bytes[i + 3] | (bytes[i + 4] << 8);
    if ((len ^ nlen) !== 0xffff) return null;
    i += 5;
    if (i + len > bytes.length) return null;
    for (let k = 0; k < len; k++) out.push(bytes[i + k]);
    i += len;
    if (header & 0x01) break;
  }
  return Uint8Array.from(out);
}

// vpnUriFromConf builds an AmneziaVPN vpn:// share link from an awg-quick
// client .conf, mirroring internal/awg/vpnuri.EncodeConf: vpn:// +
// Base64URL(qCompress(Amnezia JSON container)). AmneziaVPN's import parses
// the container natively and keeps every AWG 3.0 field (a raw-.conf import
// drops HeaderProtectionKey and friends on 5.x).
export function vpnUriFromConf(conf: string): string {
  const trimmed = conf.trim();
  if (!trimmed || !trimmed.includes('[Interface]') || !trimmed.includes('[Peer]')) return '';
  const meta = parseConf(trimmed);

  const awg: Record<string, unknown> = {
    isThirdPartyConfig: true,
    last_config: JSON.stringify(meta.inner),
    transport_proto: 'udp',
  };
  if (meta.port > 0) awg.port = String(meta.port);
  const pv = protocolVersion(meta.inner);
  if (pv) awg.protocol_version = pv;

  const env: Record<string, unknown> = {
    containers: [{ container: 'amnezia-awg', awg }],
    defaultContainer: 'amnezia-awg',
  };
  if (meta.host) env.hostName = meta.host;
  if (meta.desc) env.description = meta.desc;
  if (meta.dns1) env.dns1 = meta.dns1;
  if (meta.dns2) env.dns2 = meta.dns2;

  const payload = new TextEncoder().encode(JSON.stringify(env));
  return `vpn://${toBase64Url(qCompress(payload))}`;
}

export async function vpnConfFromLink(link: string): Promise<string> {
  const trimmed = link.trim();
  if (!trimmed.startsWith('vpn://')) return '';
  let bytes: Uint8Array;
  try {
    bytes = bytesFromBase64Url(trimmed.slice('vpn://'.length));
  } catch {
    return '';
  }
  let payload = bytes;
  if (isQCompress(bytes)) {
    try {
      payload = await inflateZlib(bytes.subarray(4));
    } catch {
      return '';
    }
  }
  return confFromPayload(payload);
}

async function inflateZlib(data: Uint8Array): Promise<Uint8Array> {
  const ds = new DecompressionStream('deflate');
  const writer = ds.writable.getWriter();
  await writer.write(data as BufferSource);
  await writer.close();
  return new Uint8Array(await new Response(ds.readable).arrayBuffer());
}

function confFromPayload(payload: Uint8Array): string {
  const trimmed = new TextDecoder().decode(payload).trim();
  if (trimmed.startsWith('{')) return confFromJson(trimmed);
  if (trimmed.includes('[Interface]') && trimmed.includes('[Peer]')) return trimmed;
  return '';
}

function confFromJson(s: string): string {
  let env: unknown;
  try {
    env = JSON.parse(s);
  } catch {
    return '';
  }
  if (!env || typeof env !== 'object') return '';
  const containers = (env as { containers?: unknown }).containers;
  if (!Array.isArray(containers) || containers.length === 0) return '';
  const last = containers[containers.length - 1];
  if (!last || typeof last !== 'object') return '';
  const rec = last as Record<string, unknown>;
  const proto = rec.awg ?? rec.wireguard;
  if (!proto || typeof proto !== 'object') return '';
  const lc = (proto as { last_config?: unknown }).last_config;
  if (typeof lc === 'string') {
    const t = lc.trim();
    if (t.startsWith('{')) {
      try {
        const inner = JSON.parse(t) as { config?: unknown };
        if (typeof inner.config === 'string' && inner.config.includes('[Interface]')) {
          return inner.config.trim();
        }
      } catch {
        /* last_config is raw .conf */
      }
    }
    if (t.includes('[Interface]')) return t;
  }
  if (lc && typeof lc === 'object') {
    const c = (lc as { config?: unknown }).config;
    if (typeof c === 'string' && c.includes('[Interface]')) return c.trim();
  }
  return '';
}
